File: src/main/java/com/tkr/desk/PositionLedger.java
package com.tkr.desk;

import com.tkr.desk.ProRataAllocator.ProRataSlice;
import com.tkr.types.WireTypes.*;
import com.tkr.types.WireTypes.Status;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Double-entry position ledger with batch apply and reconciliation.
 * Ported from legacy-cpp position_ledger.cc.
 */
public final class PositionLedger {

    public static final class PositionEntry {
        public int accountId;
        public int symbolId;
        public long qtyMilli;
        public long costBasisCents;
        public int lastTradeDate;
        public int version;
    }

    public static final class PositionLedgerConfig {
        public int deskId;

        public PositionLedgerConfig(int deskId) {
            this.deskId = deskId;
        }
    }

    public static final class PositionLedgerSnapshot {
        public int entryCount;
        public long totalQtyMilli;
        public long totalCostCents;
    }

    public static final class LedgerEntry {
        public long entryId;
        public int accountId;
        public int symbolId;
        public int debitQtyMilli;
        public int creditQtyMilli;
        public long timestampMillis;
        public String memo = "";
    }

    public static final class AccountBalance {
        public int accountId;
        public int symbolId;
        public int netQtyMilli;
    }

    private static final class PositionDelta {
        int accountId;
        int symbolId;
        long qtyDelta;
        long costDelta;
    }

    private final PositionLedgerConfig config;
    private final List<PositionEntry> entries = new ArrayList<>();
    private final Map<Long, Integer> index = new HashMap<>();
    private long nextEntryId = 1;
    private final List<LedgerEntry> journal = new ArrayList<>();
    private final Map<Long, AccountBalance> balances = new HashMap<>();

    public PositionLedger() {
        this(new PositionLedgerConfig(0));
    }

    public PositionLedger(PositionLedgerConfig config) {
        this.config = config != null ? config : new PositionLedgerConfig(0);
    }

    private static long makeKey(int accountId, int symbolId) {
        return ((long) accountId << 32) | (symbolId & 0xFFFFFFFFL);
    }

    private long balanceKey(int accountId, int symbolId) {
        return makeKey(accountId, symbolId);
    }

    public Status validateSlice(ProRataSlice slice) {
        if (slice == null || slice.accountId == 0) return Status.BOUNDS_ERROR;
        return Status.OK;
    }

    public Status upsertEntry(int accountId, int symbolId, long deltaQty, long deltaCost, int tradeDate) {
        long key = makeKey(accountId, symbolId);
        Integer idx = index.get(key);
        if (idx != null) {
            PositionEntry entry = entries.get(idx);
            entry.qtyMilli += deltaQty;
            entry.costBasisCents += deltaCost;
            entry.lastTradeDate = tradeDate;
            entry.version++;
            return Status.OK;
        }
        PositionEntry entry = new PositionEntry();
        entry.accountId = accountId;
        entry.symbolId = symbolId;
        entry.qtyMilli = deltaQty;
        entry.costBasisCents = deltaCost;
        entry.lastTradeDate = tradeDate;
        entry.version = 1;
        index.put(key, entries.size());
        entries.add(entry);
        return Status.OK;
    }

    public Status applySlice(ProRataSlice slice, int symbolId, int priceTick) {
        Status valid = validateSlice(slice);
        if (valid != Status.OK) return valid;
        long cost = (long) slice.qtyMilli * priceTick / 1000L;
        return upsertEntry(slice.accountId, symbolId, slice.qtyMilli, cost, 0);
    }

    public Status applyBatch(BatchWireFrame frame, List<ProRataSlice> slices) {
        if (slices.isEmpty() && frame.records.isEmpty()) return Status.OK;
        if (slices.size() != frame.records.size()) return Status.BOUNDS_ERROR;
        List<PositionDelta> deltas = new ArrayList<>();
        for (int i = 0; i < slices.size(); i++) {
            ProRataSlice slice = slices.get(i);
            WireBatchRecord rec = frame.records.get(i);
            PositionDelta delta = new PositionDelta();
            delta.accountId = slice.accountId;
            delta.symbolId = rec.symbolId;
            delta.qtyDelta = slice.qtyMilli;
            delta.costDelta = (long) slice.qtyMilli * rec.priceTick / 1000L;
            deltas.add(delta);
        }
        deltas.sort(Comparator.comparingInt((PositionDelta d) -> d.accountId).thenComparingInt(d -> d.symbolId));
        for (PositionDelta delta : deltas) {
            Status st = upsertEntry(delta.accountId, delta.symbolId, delta.qtyDelta, delta.costDelta,
                    frame.header.tradeDateYyyymmdd);
            if (st != Status.OK) return st;
        }
        return Status.OK;
    }

    public PositionEntry lookup(int accountId, int symbolId) {
        Integer idx = index.get(makeKey(accountId, symbolId));
        return idx != null ? entries.get(idx) : new PositionEntry();
    }

    public PositionLedgerSnapshot snapshot() {
        PositionLedgerSnapshot snap = new PositionLedgerSnapshot();
        snap.entryCount = entries.size();
        for (PositionEntry entry : entries) {
            snap.totalQtyMilli += entry.qtyMilli;
            snap.totalCostCents += entry.costBasisCents;
        }
        return snap;
    }

    public void reset() {
        entries.clear();
        index.clear();
        journal.clear();
        balances.clear();
    }

    public List<PositionEntry> entriesByAccount(int accountId) {
        List<PositionEntry> result = new ArrayList<>();
        for (PositionEntry entry : entries) {
            if (entry.accountId == accountId) result.add(entry);
        }
        return result;
    }

    public long totalQtyForSymbol(int symbolId) {
        long total = 0;
        for (PositionEntry entry : entries) {
            if (entry.symbolId == symbolId) total += entry.qtyMilli;
        }
        return total;
    }

    public Status reconcileEntry(int accountId, int symbolId, long expectedQty) {
        PositionEntry entry = lookup(accountId, symbolId);
        if (entry.accountId == 0) return Status.BOUNDS_ERROR;
        return entry.qtyMilli == expectedQty ? Status.OK : Status.BOUNDS_ERROR;
    }

    public Status mergeEntries(PositionEntry other) {
        return upsertEntry(other.accountId, other.symbolId, other.qtyMilli, other.costBasisCents, other.lastTradeDate);
    }

    public boolean hasOpenPosition(int accountId, int symbolId) {
        return lookup(accountId, symbolId).qtyMilli != 0;
    }

    public long averageCost(int accountId, int symbolId) {
        PositionEntry entry = lookup(accountId, symbolId);
        return entry.qtyMilli == 0 ? 0 : entry.costBasisCents / entry.qtyMilli;
    }

    /** Legacy double-entry journal API. */
    public void postAllocation(int accountId, int qtyMilli, int symbolId) {
        postPair(accountId, symbolId, qtyMilli, 0, "allocation");
        postPair(0, symbolId, 0, qtyMilli, "allocation_contra");
    }

    public void postPair(int accountId, int symbolId, int debit, int credit, String memo) {
        LedgerEntry entry = new LedgerEntry();
        entry.entryId = nextEntryId++;
        entry.accountId = accountId;
        entry.symbolId = symbolId;
        entry.debitQtyMilli = debit;
        entry.creditQtyMilli = credit;
        entry.timestampMillis = System.currentTimeMillis();
        entry.memo = memo;
        journal.add(entry);
        AccountBalance bal = balances.computeIfAbsent(balanceKey(accountId, symbolId), k -> {
            AccountBalance b = new AccountBalance();
            b.accountId = accountId;
            b.symbolId = symbolId;
            return b;
        });
        bal.netQtyMilli += debit - credit;
    }

    public int netQty(int accountId, int symbolId) {
        AccountBalance bal = balances.get(balanceKey(accountId, symbolId));
        return bal != null ? bal.netQtyMilli : 0;
    }

    public Status verifyBalanced(int symbolId) {
        int net = 0;
        for (AccountBalance b : balances.values()) {
            if (b.symbolId == symbolId) net += b.netQtyMilli;
        }
        return net == 0 ? Status.OK : Status.BOUNDS_ERROR;
    }

    public List<LedgerEntry> snapshotJournal() {
        return new ArrayList<>(journal);
    }

    public int totalDebitQty(int symbolId) {
        int sum = 0;
        for (LedgerEntry e : journal) if (e.symbolId == symbolId) sum += e.debitQtyMilli;
        return sum;
    }

    public int totalCreditQty(int symbolId) {
        int sum = 0;
        for (LedgerEntry e : journal) if (e.symbolId == symbolId) sum += e.creditQtyMilli;
        return sum;
    }

    public List<PositionEntry> allEntries() {
        return new ArrayList<>(entries);
    }
}

File: legacy-cpp/src/desk/position_ledger.cc
#include "tkr/desk/position_ledger.h"

#include <algorithm>

namespace tkr {
namespace desk {
namespace {

struct PositionDelta {
  std::uint32_t account_id;
  std::uint32_t symbol_id;
  std::int64_t qty_delta;
  std::int64_t cost_delta;
};

struct LedgerAuditEntry {
  std::uint32_t account_id;
  std::uint32_t symbol_id;
  std::int64_t qty_before;
  std::int64_t qty_after;
  std::uint32_t version;
};

}  // namespace

PositionLedger::PositionLedger(PositionLedgerConfig config) : config_(config) {}

std::uint64_t PositionLedger::MakeKey(std::uint32_t account_id,
                                      std::uint32_t symbol_id) const {
  return (static_cast<std::uint64_t>(account_id) << 32) |
         static_cast<std::uint64_t>(symbol_id);
}

Status PositionLedger::ValidateSlice(const ProRataSlice& slice) const {
  if (slice.account_id == 0) {
    return Status::kBoundsError;
  }
  return Status::kOk;
}

Status PositionLedger::UpsertEntry(std::uint32_t account_id,
                                   std::uint32_t symbol_id,
                                   std::int64_t delta_qty,
                                   std::int64_t delta_cost,
                                   std::uint32_t trade_date) {
  const std::uint64_t key = MakeKey(account_id, symbol_id);
  const auto it = index_.find(key);

  if (it != index_.end()) {
    PositionEntry& entry = entries_[it->second];
    entry.qty_milli += delta_qty;
    entry.cost_basis_cents += delta_cost;
    entry.last_trade_date = trade_date;
    entry.version += 1;
    return Status::kOk;
  }

  PositionEntry entry{};
  entry.account_id = account_id;
  entry.symbol_id = symbol_id;
  entry.qty_milli = delta_qty;
  entry.cost_basis_cents = delta_cost;
  entry.last_trade_date = trade_date;
  entry.version = 1;

  index_[key] = entries_.size();
  entries_.push_back(entry);
  return Status::kOk;
}

Status PositionLedger::ApplySlice(const ProRataSlice& slice,
                                  std::uint32_t symbol_id,
                                  std::uint32_t price_tick) {
  Status valid = ValidateSlice(slice);
  if (valid != Status::kOk) {
    return valid;
  }

  const std::int64_t cost =
      (static_cast<std::int64_t>(slice.qty_milli) *
       static_cast<std::int64_t>(price_tick)) /
      1000;

  return UpsertEntry(slice.account_id, symbol_id,
                     static_cast<std::int64_t>(slice.qty_milli), cost, 0);
}

Status PositionLedger::ApplyBatch(
    const BatchWireFrame& frame, const std::vector<ProRataSlice>& slices) {
  if (slices.empty() && frame.records.empty()) {
    return Status::kOk;
  }

  if (slices.size() != frame.records.size()) {
    return Status::kBoundsError;
  }

  std::vector<PositionDelta> deltas;
  deltas.reserve(slices.size());

  for (std::size_t i = 0; i < slices.size(); ++i) {
    const ProRataSlice& slice = slices[i];
    const WireBatchRecord& rec = frame.records[i];

    PositionDelta delta{};
    delta.account_id = slice.account_id;
    delta.symbol_id = rec.symbol_id;
    delta.qty_delta = static_cast<std::int64_t>(slice.qty_milli);
    delta.cost_delta =
        (static_cast<std::int64_t>(slice.qty_milli) *
         static_cast<std::int64_t>(rec.price_tick)) /
        1000;
    deltas.push_back(delta);
  }

  std::sort(deltas.begin(), deltas.end(),
            [](const PositionDelta& a, const PositionDelta& b) {
              if (a.account_id != b.account_id) {
                return a.account_id < b.account_id;
              }
              return a.symbol_id < b.symbol_id;
            });

  for (const PositionDelta& delta : deltas) {
    Status st = UpsertEntry(delta.account_id, delta.symbol_id, delta.qty_delta,
                            delta.cost_delta, frame.header.trade_date_yyyymmdd);
    if (st != Status::kOk) {
      return st;
    }
  }

  return Status::kOk;
}

PositionEntry PositionLedger::Lookup(std::uint32_t account_id,
                                     std::uint32_t symbol_id) const {
  const std::uint64_t key = MakeKey(account_id, symbol_id);
  const auto it = index_.find(key);
  if (it == index_.end()) {
    return PositionEntry{};
  }
  return entries_[it->second];
}

PositionLedgerSnapshot PositionLedger::Snapshot() const {
  PositionLedgerSnapshot snap{};
  snap.entry_count = static_cast<std::uint32_t>(entries_.size());
  for (const PositionEntry& entry : entries_) {
    snap.total_qty_milli += entry.qty_milli;
    snap.total_cost_cents += entry.cost_basis_cents;
  }
  return snap;
}

void PositionLedger::Reset() {
  entries_.clear();
  index_.clear();
}

std::vector<PositionEntry> PositionLedger::EntriesByAccount(
    std::uint32_t account_id) const {
  std::vector<PositionEntry> result;
  for (const PositionEntry& entry : entries_) {
    if (entry.account_id == account_id) {
      result.push_back(entry);
    }
  }
  return result;
}

std::int64_t PositionLedger::TotalQtyForSymbol(std::uint32_t symbol_id) const {
  std::int64_t total = 0;
  for (const PositionEntry& entry : entries_) {
    if (entry.symbol_id == symbol_id) {
      total += entry.qty_milli;
    }
  }
  return total;
}

Status PositionLedger::ReconcileEntry(std::uint32_t account_id,
                                      std::uint32_t symbol_id,
                                      std::int64_t expected_qty) {
  PositionEntry entry = Lookup(account_id, symbol_id);
  if (entry.account_id == 0) {
    return Status::kBoundsError;
  }
  if (entry.qty_milli != expected_qty) {
    return Status::kBoundsError;
  }
  return Status::kOk;
}

Status PositionLedger::MergeEntries(const PositionEntry& other) {
  return UpsertEntry(other.account_id, other.symbol_id, other.qty_milli,
                     other.cost_basis_cents, other.last_trade_date);
}

bool PositionLedger::HasOpenPosition(std::uint32_t account_id,
                                     std::uint32_t symbol_id) const {
  const PositionEntry entry = Lookup(account_id, symbol_id);
  return entry.qty_milli != 0;
}

std::int64_t PositionLedger::AverageCost(std::uint32_t account_id,
                                         std::uint32_t symbol_id) const {
  const PositionEntry entry = Lookup(account_id, symbol_id);
  if (entry.qty_milli == 0) {
    return 0;
  }
  return entry.cost_basis_cents / entry.qty_milli;
}

}  // namespace desk
}  // namespace tkr
